### PR TITLE
Explicitly declare that factories support `array<mixed>` for construct-time options, instead of just declaring `array`

### PR DESCRIPTION
**Pull request by fezfez**
see https://github.com/Slamdunk/phpstan-laminas-framework/issues/26

### CODE REVIEW

File: src/Factory/FactoryInterface.php [hunk 1]
```diff
@@ -23,7 +23,7 @@ interface FactoryInterface
      * Create an object
      *
      * @param  string             $requestedName
-     * @param  null|array         $options
+     * @param  null|array<mixed>  $options
```
Comment (Ocramius) [inline src/Factory/FactoryInterface.php:26]: `array` is an alias of `array<int|string, mixed>`? Why is this needed explicitly? :thinking:

**Review by Ocramius — CHANGES_REQUESTED** [review]

**Review by Ocramius — APPROVED** [review]
Hmm, I see https://github.com/laminas/laminas-servicemanager/pull/99#discussion_r680998436 is about strictness: from my PoV it's OK to declare it `array<int|string, mixed>`, but it seems weird that this is causing issues for subtype/implementations.

**Comment by Ocramius** [comment]
Thanks @fezfez!

File: src/Factory/FactoryInterface.php (hunk 1 above)
Comment (fezfez) [reply src/Factory/FactoryInterface.php:26]: Yeah, that's what i was thinking. Maybe i should report to phpstan instead.

**Comment by Ocramius** [comment]
For others wondering whether this is necessary at all: it isn't, but you can always go back from `array<mixed>` to just `array`, so it doesn't hurt anyway.

**Comment by boesing** [comment]
Actually, `options` should be `array<string,mixed>` as options supposed to be a map.
Thats what I've added in #98 btw.

**Comment by Ocramius** [comment]
@boesing is it really restricted to a map? :thinking:

**Comment by boesing** [comment]
@Ocramius its not but just because it is not declared at all and nobody cares.

Options are mostly passed to plugins where they are either stored by using setters on the same plugin (e.g. validators, filters) or in case of the cache component, passed to the `AdapterOptions`. It just dont make any sense to allow integer keys for an Options Argument...

**Comment by Ocramius** [comment]
@boesing I would expect this to be valid though:

```php
$sevenDwarvesFacility = $pluginManager->create(MiningFacility::class, [$dopey, $grumpy, $bashful, $sleepy, $sneezy]);
```

**Comment by boesing** [comment]
@Ocramius so you expect seven dwarves for 5 variables passed as options? 🤪

I see the point but I would expect passing these variables as an Option "dwarves". 
But whatever, gonna update my PR - has conflicts now anyways.

I would prefer letting the plugin manager to declare what type is expected. Gonna try something with generics, maybe I get something that might work.

**Comment by Ocramius** [comment]
I was just making an example of something that I consider valid: I rarely use options for service creation, but I don't see big value in restricting it in this location, as it's just a mechanism for parameter passthrough, and we'd only introduce a variance BC break.
